gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -0,0 +1,288 @@
+"""Inference methods for GtsfmData in PatchmatchNet
+
+Authors: Ren Liu
+"""
+import time
+from typing import Dict, List
+
+import numpy as np
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader
+
+import gtsfm.densify.mvs_utils as mvs_utils
+from gtsfm.common.gtsfm_data import GtsfmData
+from gtsfm.common.image import Image
+from gtsfm.densify.mvs_base import MVSBase
+from gtsfm.densify.patchmatchnet_data import PatchmatchNetData
+from gtsfm.utils import logger as logger_utils
+import thirdparty.patchmatchnet.utils as patchmatchnet_utils
+import thirdparty.patchmatchnet.eval as patchmatchnet_eval
+from thirdparty.patchmatchnet.models.net import PatchmatchNet
+
+logger = logger_utils.get_logger()
+
+torch.backends.cudnn.benchmark = True
+PATCHMATCHNET_WEIGHTS_PATH = "thirdparty/patchmatchnet/checkpoints/model_000007.ckpt"
+
+# all default values are assigned by Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/eval.py
+DEFAULT_BATCH_SIZE = 1
+DEFAULT_NUM_VIEWS = 5
+DEFAULT_NUM_WORKERS = 4
+
+# default pixel coordinates error threshold for geometric consistency check
+DEFAULT_GEOMETRIC_PIXEL_THRESH = 1.0
+# default depth error threshold for geometric consistency check
+DEFAULT_GEOMETRIC_DEPTH_THRESH = 0.01
+DEFAULT_CONFIDENCE_THRESH = 0.8
+
+DEFAULT_INTERVAL_SCALE = [0.005, 0.0125, 0.025]
+DEFAULT_PROPAGATION_RANGE = [6, 4, 2]
+DEFAULT_NUM_ITERS = [1, 2, 2]
+DEFAULT_NUM_SAMPLES = [8, 8, 16]
+DEFAULT_PROPAGATE_NEIGHBORS = [0, 8, 16]
+DEFAULT_EVALUATE_NEIGHBORS = [9, 9, 9]
+
+# a reconstructed point is consistent in geometry if it satisfies all gemetric thresholds in more than 3 source views
+DEFAULT_MIN_NUM_CONSISTENT_VIEWS = 3
+
+
+class MVSPatchmatchNet(MVSBase):
+    """Inference methods for GtsfmData in densification using PatchmatchNet."""
+
+    def densify(
+        self,
+        images: Dict[int, Image],
+        sfm_result: GtsfmData,
+        max_num_views: int = DEFAULT_NUM_VIEWS,
+        thresholds: List[float] = [
+            DEFAULT_GEOMETRIC_PIXEL_THRESH,
+            DEFAULT_GEOMETRIC_DEPTH_THRESH,
+            DEFAULT_CONFIDENCE_THRESH,
+        ],
+    ) -> np.ndarray:
+        """Get dense point cloud using PatchmatchNet from GtsfmData. The method implements the densify method in MVSBase
+        Ref: Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/eval.py
+
+        Args:
+            images: image dictionary obtained from loaders
+            sfm_result: result of GTSFM after bundle adjustment
+            num_views: maximum number of views, containing 1 reference view and (num_views-1) source views
```

Comment:
i think this should be `max_num_views` to be consistent with function signature @codyly ?